src/xrpld/app/tx/detail/VaultClawback.cpp

```diff
@@ -47,51 +43,74 @@ VaultClawback::preclaim(PreclaimContext const& ctx)
+    auto const sleShareIssuance =
+        ctx.view.read(keylet::mptIssuance(mptIssuanceID));
+    if (!sleShareIssuance)
     {
         // LCOV_EXCL_START
-        JLOG(ctx.j.error()) << "VaultClawback: missing issuer account.";
+        JLOG(ctx.j.error())
+            << "VaultClawback: missing issuance of vault shares.";
         return tefINTERNAL;
         // LCOV_EXCL_STOP
     }
 
-    Asset const vaultAsset = vault->at(sfAsset);
-    if (auto const amount = ctx.tx[~sfAmount];
-        amount && vaultAsset != amount->asset())
+    Asset const share = MPTIssue{mptIssuanceID};
+    // Ambiguous case: If the Issuer is the Vault Owner the must explicitly
+    // provide the asset
+    if (!maybeAmount && !vaultAsset.native() &&
+        vaultAsset.getIssuer() == vault->at(sfOwner))
+    {
+        JLOG(ctx.j.debug())
+            << "VaultClawback: must specify amount when issuer is owner.";
         return tecWRONG_ASSET;
+    }
+
+    auto const amount = [&]() {
+        if (maybeAmount)
+            return *maybeAmount;
+
+        if (account == vault->at(sfOwner))
+            return STAmount{share};
+
+        return STAmount{vaultAsset};
+    }();
 
     // There is a special case that allows the VaultOwner to use clawback to
     // burn shares when Vault assets total and available are zero, but
     // shares remain. However, that case is handled in doApply() directly,
     // so here we just enforce checks.
-    auto const mptIssuanceID = vault->at(sfShareMPTID);
-    auto const sleShareIssuance =
-        ctx.view.read(keylet::mptIssuance(mptIssuanceID));
-    if (!sleShareIssuance)
+    if (amount.asset() == share)
     {
-        // LCOV_EXCL_START
-        JLOG(ctx.j.error())
-            << "VaultClawback: missing issuance of vault shares.";
-        return tefINTERNAL;
-        // LCOV_EXCL_STOP
-    }
+        // Only the Vault Owner may clawback shares
+        if (account != vault->at(sfOwner))
+        {
+            JLOG(ctx.j.debug())
+                << "VaultClawback: only vault owner can clawback shares.";
+            return tecNO_PERMISSION;
+        }
 
-    auto const assetsTotal = vault->at(sfAssetsTotal);
-    auto const assetsAvailable = vault->at(sfAssetsAvailable);
-    auto const sharesTotal = sleShareIssuance->at(sfOutstandingAmount);
-    auto const owner = vault->at(sfOwner);
-    auto const share = MPTIssue{mptIssuanceID};
+        auto const assetsTotal = vault->at(sfAssetsTotal);
+        auto const assetsAvailable = vault->at(sfAssetsAvailable);
+        auto const sharesTotal = sleShareIssuance->at(sfOutstandingAmount);
 
-    // Allow clawback to burn shares in this special case.
-    if (sharesTotal > 0 && assetsTotal == 0 && assetsAvailable == 0 &&
-        account == owner)
-    {
-        if (auto const amount = ctx.tx[~sfAmount]; amount)
+        // Owner can clawback funds when the vault has shares but no assets
+        if (sharesTotal != 0 && (assetsTotal != 0 || assetsAvailable != 0))
```

Comment:
I think this should fail when `sharesTotal == 0`, too. You could add a separate case, or you could change this condition slightly
```suggestion
        if (sharesTotal == 0 || (assetsTotal != 0 || assetsAvailable != 0))
```
And then change the log message to "VaultClawback: vault owner can clawback shares only when vault has shares, but no assets."